lte/gateway/c/core/oai/tasks/async_grpc_service/S6aGatewayImpl.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <thread>
+#include "includes/MagmaService.h"
+#include "S6aGatewayImpl.h"
+
+extern "C" {
+#include "log.h"
+#include "s6a_service_handler.h"
+}
+
+namespace magma {
+#define S6A_ASYNC_PROXY_SERVICE "s6a_async_service"
+#define S6A_ASYNC_PROXY_VERSION "1.0"
+
+magma::S6aProxyResponderAsyncService s6a_async_service(nullptr, nullptr);
+static magma::service303::MagmaService server(
+    S6A_ASYNC_PROXY_SERVICE, S6A_ASYNC_PROXY_VERSION);
+
+void stop_async_s6a_grpc_server(void) {
+  s6a_async_service.stop();  // stop queue after server shuts down
+}
+
+void init_async_s6a_grpc_server(void) {
+  auto async_service_handler =
+      std::make_shared<magma::S6aProxyAsyncResponderHandler>();
+  s6a_async_service.set_callback(
+      server.GetNewCompletionQueue(), async_service_handler);
+  server.AddServiceToServer(&s6a_async_service);
+
+  server.Start();
+  OAILOG_INFO(LOG_S6A, "Started async grpc server for s6a interface \n");
+
+  std::thread proxy_thread([&]() {
+    s6a_async_service.wait_for_requests();  // block here instead of on server
+  });
+  proxy_thread.join();
```

Comment:
if you do this, you are waiting for proxy_thread to end and will not be able to start the zloop thread. Remove the join line. Also do not use a global in c++ land, if you really need to have a single instance, utilize singleton class.